### PR TITLE
Prac 05 feedback

### PR DESCRIPTION
**Pull request by JarredSnow**

### CODE REVIEW

**Comment by JarredSnow** [comment]
@Dylan-Perkins04 please provide a code review for CP1404

File: prac_05/state_names.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+"""
+CP1404/CP5632 Practical
+State names in a dictionary
+File needs reformatting
+"""
```
Comment (Dylan-Perkins04) [inline prac_05/state_names.py]: It's good that you have a doc string here, however I believe this is the old docstring from the template file Lindsay provided us. I recommend removing the old information such as "File needs reformatting" and get rid of the "/CP5632"

File: prac_05/state_names.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+MAX_KEY_LEN = len(max(CODE_TO_NAME, key=len))
+MIN_KEY_LEN = len(min(CODE_TO_NAME, key=len))
+MAX_VALUE_LEN = len(CODE_TO_NAME[max(CODE_TO_NAME, key=len)])
+MIN_VALUE_LEN = len(CODE_TO_NAME[min(CODE_TO_NAME, key=len)])
+
+
+
+def main():
+    """Turn short state codes to full names, then prints the entire key"""
```
Comment (Dylan-Perkins04) [inline prac_05/state_names.py:23]: This doc string seems a bit wordy. Could possibly be changed to something like "Convert state codes to state names".
Also the main function doesn't only print the "entire key", instead it prints out the dictionary.

File: prac_05/state_names.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    decode_short_state()
+    print_state(CODE_TO_NAME)
+
+
+def decode_short_state():
```
Comment (Dylan-Perkins04) [inline prac_05/state_names.py:28]: Could the function name be better? Try saying out loud "This function will decode short state", does it flow well?

File: prac_05/state_names.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    """Turn short state codes to full names"""
+    while True:
```
Comment (Dylan-Perkins04) [inline prac_05/state_names.py:30]: Have we ever used while True loops in this subject?
Have a look at the error checking loop pattern available on learn jcu.

File: prac_05/state_names.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            print(state_code, "is", CODE_TO_NAME[state_code])
+            break
+        except KeyError:
+            print("Invalid short state")
+
+
+def print_state(state_codes):
+    """Print short state key for changing to full name"""
+    for state in state_codes:
+        print(f"{state:<{MAX_KEY_LEN}} is {state_codes[state]:<{MAX_VALUE_LEN}}")
```
Comment (Dylan-Perkins04) [inline prac_05/state_names.py:45]: Do you think we need a seperate function just for a "for loop"?
Could possibly be put in main instead

File: prac_05/state_names.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                "NT": "Northern Territory",
+                "WA": "Western Australia",
+                "ACT": "Australian Capital Territory",
+                "VIC": "Victoria",
+                "TAS": "Tasmania"}
+
+MAX_KEY_LEN = len(max(CODE_TO_NAME, key=len))
+MIN_KEY_LEN = len(min(CODE_TO_NAME, key=len))
+MAX_VALUE_LEN = len(CODE_TO_NAME[max(CODE_TO_NAME, key=len)])
+MIN_VALUE_LEN = len(CODE_TO_NAME[min(CODE_TO_NAME, key=len)])
```
Comment (Dylan-Perkins04) [inline prac_05/state_names.py:18]: When I first saw these constants I was a bit confused and overwhelmed, are all of these constants needed?
We already know what the max length of the keys are... 3.
Also you didn't use either of the MIN constants, redundant code.

**Review by Dylan-Perkins04 — COMMENTED** [review]
Most of the issues were involving your functions.
Do you think you need functions for this program? The template code didn't use functions and using them has caused quite a few issues.


Did you struggle naming the functions and writing the docstrings for them? If you did, maybe that's a sign that you don't need to use them

File: prac_05/hex_colours.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+"""
+CP1404/CP5632 Practical
```
Comment (Dylan-Perkins04) [inline prac_05/hex_colours.py]: Good docstring :)
However are you doing both CP1404 and CP5632?
I would change it to (your subject) Practical

File: prac_05/hex_colours.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+Hex Colours Dictionary Lookup
+"""
+
+colour_codes = {
```
Comment (Dylan-Perkins04) [inline prac_05/hex_colours.py:6]: Does this follow the naming convention for dictionaries?
Have a look through the week 5 ppt slides

File: prac_05/hex_colours.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        'daffodil': '#ffff31',
+        'ebony': '555d50',
+        'emerald': '#50c878',
+        'firebrick': '#ff3030',
+        'gray': '#bebebe',
+        'heliotrope': '#df73ff',
+        'magenta': '#ff00ff'
+        }
+
+while True:
```
Comment (Dylan-Perkins04) [inline prac_05/hex_colours.py:19]: Another while True loop... have we ever used this kind of loop?

Look at the error checking loop pattern, should look something like while colour != ""

File: prac_05/hex_colours.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    try:
+        colour_choice = input('Select your colour:').lower()
+        if colour_choice == '':
+            break
+        print(f'The code for {colour_choice} is {colour_codes[colour_choice]}')
+        break
```
Comment (Dylan-Perkins04) [inline prac_05/hex_colours.py:25]: How come you used single quote marks for this print statement, but used double quote marks for the last print statement?

Also I feel like the break here will cause the loop to only allow the user to input one colour...

**Review by Dylan-Perkins04 — COMMENTED** [review]
A few little issues, but using a while True loop is pretty concerning. Did you use ChatGPT for this?

**Comment by JarredSnow** [comment]
> Most of the issues were involving your functions. Do you think you need functions for this program? The template code didn't use functions and using them has caused quite a few issues.
> 
> Did you struggle naming the functions and writing the docstrings for them? If you did, maybe that's a sign that you don't need to use them

I appreciate you taking a look at it, that's good advice I will try to keep it in mind.

File: prac_05/state_names.py (hunk 1 above)
Comment (JarredSnow) [reply prac_05/state_names.py]: Thank you I appreciate you pointing that out.

File: prac_05/state_names.py (hunk 2 above)
Comment (JarredSnow) [reply prac_05/state_names.py:23]: Thank you, I do struggle with writing doc strings.

File: prac_05/state_names.py (hunk 3 above)
Comment (JarredSnow) [reply prac_05/state_names.py:28]: Yeah I do need to practice better naming appreciate it.

File: prac_05/state_names.py (hunk 4 above)
Comment (JarredSnow) [reply prac_05/state_names.py:30]: I don't think we have, this is just something I do form other python stuff I have done. I will have to look at how Lindsay teaches it.

File: prac_05/state_names.py (hunk 5 above)
Comment (JarredSnow) [reply prac_05/state_names.py:45]: Probably not, I am mostly trying to practice using them but should consider that thank you.

File: prac_05/state_names.py (hunk 6 above)
Comment (JarredSnow) [reply prac_05/state_names.py:18]: So with this it's for practice using these types of lines, I also think it is maybe more robust than using an integer(what if we decide to add something that isn't length 3). I guess with the context of the problem 3 is probably more appropriate though. MIN was just more practice and just for future use again. Probably inappropriate for the problem as set.

**Comment by JarredSnow** [comment]
> A few little issues, but using a while True loop is pretty concerning. Did you use ChatGPT for this?

No I haven't used ChatGPT before for code, it's just something I learned from self teaching python from text books/stack overflow examplesprior to the subject.

File: prac_05/hex_colours.py (hunk 7 above)
Comment (JarredSnow) [reply prac_05/hex_colours.py]: I kept the CP5 stuff because it was in the provided examples.

**Comment by Dylan-Perkins04** [comment]
Thats all good 🙂
I just finished reviewing everything and I hope it helps
________________________________

**Comment by JarredSnow** [comment]
> Thats all good 🙂 I just finished reviewing everything and I hope it helps
> […](#)

definitely does your a legend.
